File: CursoAvanzadodeServerlessFrameworkenAWS/pruebas/handler.js
const AWS = require("aws-sdk");
const { v4: uuidv4 } = require("uuid");

let dynamoDBClientParams = {};

if (process.env.IS_OFFLINE === 'true') {
  dynamoDBClientParams = {
    region: 'localhost',
    endpoint: 'http://localhost:8000',
    accessKeyId: 'DEFAULT_ACCESS_KEY',
    secretAccessKey: 'DEFAULT_SECRET'
  };
}

const dynamodb = new AWS.DynamoDB.DocumentClient(dynamoDBClientParams);
const TABLE_NAME = process.env.USERS_TABLE;

module.exports.handler = async (event) => {
  const method = event.httpMethod;
  const id = event.pathParameters?.id;

  switch (method) {
    case 'GET':
      return await getUser(id);
    case 'POST':
      return await createUser(JSON.parse(event.body));
    case 'PUT':
      return await updateUser(id, JSON.parse(event.body));
    case 'DELETE':
      return await deleteUser(id);
    default:
      return { statusCode: 400, body: "Método no soportado" };
  }
};

// Funciones CRUD
const getUser = async (id) => {
  const res = await dynamodb.get({
    TableName: TABLE_NAME,
    Key: { pk: id }
  }).promise();

  if (!res.Item) {
    return { statusCode: 404, body: JSON.stringify({ error: "Usuario no encontrado" }) };
  }

  return { statusCode: 200, body: JSON.stringify(res.Item) };
};

const createUser = async (data) => {
  const id = uuidv4();
  const item = { pk: id, ...data };

  await dynamodb.put({
    TableName: TABLE_NAME,
    Item: item
  }).promise();

  return { statusCode: 201, body: JSON.stringify(item) };
};

const updateUser = async (id, data) => {
  const params = {
    TableName: TABLE_NAME,
    Key: { pk: id },
    UpdateExpression: "set #name = :name, #age = :age",
    ExpressionAttributeNames: {
      "#name": "name",
      "#age": "age"
    },
    ExpressionAttributeValues: {
      ":name": data.name,
      ":age": data.age
    },
    ReturnValues: "ALL_NEW"
  };

  const res = await dynamodb.update(params).promise();
  return { statusCode: 200, body: JSON.stringify(res.Attributes) };
};

const deleteUser = async (id) => {
  await dynamodb.delete({
    TableName: TABLE_NAME,
    Key: { pk: id }
  }).promise();

  return { statusCode: 200, body: JSON.stringify({ mensaje: "Usuario eliminado" }) };
};

File: CursoAvanzadodeServerlessFrameworkenAWS/pruebaenpyton/handler.py
import os
import json
import uuid
import boto3

is_offline = os.environ.get("IS_OFFLINE") == "true"

dynamodb = boto3.resource(
    'dynamodb',
    endpoint_url="http://localhost:8000" if is_offline else None,
    region_name="us-east-1",
    aws_access_key_id="DEFAULT_ACCESS_KEY" if is_offline else None,
    aws_secret_access_key="DEFAULT_SECRET" if is_offline else None
)

table = dynamodb.Table(os.environ.get("USERS_TABLE"))

def handler(event, context):
    http_method = event['httpMethod']
    user_id = (event.get("pathParameters") or {}).get("id")

    if http_method == "GET":
        return get_user(user_id)
    elif http_method == "POST":
        return create_user(json.loads(event["body"]))
    elif http_method == "PUT":
        return update_user(user_id, json.loads(event["body"]))
    elif http_method == "DELETE":
        return delete_user(user_id)
    else:
        return response(400, {"error": "Método no soportado"})

def get_user(user_id):
    try:
        res = table.get_item(Key={"pk": user_id})
        if "Item" not in res:
            return response(404, {"error": "Usuario no encontrado"})
        return response(200, res["Item"])
    except Exception as e:
        return response(500, {"error": str(e)})

def create_user(data):
    try:
        new_id = str(uuid.uuid4())
        item = {"pk": new_id, **data}
        table.put_item(Item=item)
        return response(201, item)
    except Exception as e:
        return response(500, {"error": str(e)})

def update_user(user_id, data):
    try:
        res = table.update_item(
            Key={"pk": user_id},
            UpdateExpression="SET #n = :name, #a = :age",
            ExpressionAttributeNames={
                "#n": "name",
                "#a": "age"
            },
            ExpressionAttributeValues={
                ":name": data["name"],
                ":age": data["age"]
            },
            ReturnValues="ALL_NEW"
        )
        return response(200, res["Attributes"])
    except Exception as e:
        return response(500, {"error": str(e)})

def delete_user(user_id):
    try:
        table.delete_item(Key={"pk": user_id})
        return response(200, {"message": "Usuario eliminado"})
    except Exception as e:
        return response(500, {"error": str(e)})

def response(status_code, body):
    return {
        "statusCode": status_code,
        "body": json.dumps(body),
        "headers": {
            "Content-Type": "application/json"
        }
    }
